### PR TITLE
makefiles/blobs: Add rules for binary blob embedding. 

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

This PR adds a new makefile module for embedding raw binary data intoexecutables. These is useful for including resources such as graphics (for UIs), certificates, crypto keys, scripts, etc.

Previously such a global rule did not exist and people had to resort to either provide a pre-generated C-file with an array or hand-craft a (header) file with xxd. The advantage of the new rules are:

- No additional tools required (uses only the toolchain's tools)
- Declared data types are easy to configure (no sed magic.)
- No arrays in headers.
- Output section is configurable (this makes it possible to include code, not only data.)

### Testing procedure

I migrated the two Lua examples to the new system. They should compile and run without problems.

### Issues/PRs references

Partial alternative to #9565.
See, for example, this file: https://github.com/RIOT-OS/RIOT/pull/10308/files#diff-09ce13303248a398e36e3febb371129a

### CODE REVIEW

**Comment by kaspar030** [comment]
Thanks for working on this!

> * No arrays in headers.

Whats wrong with that? Its at least industry best practice, and the performance overhead is negligible.

File: makefiles/blobs.inc.mk [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+# Generate a header with extern declarations to access the blob contents.
+# Parameters:
+#	$1: blob_file_id
+#	$2: data_type
+#	$3: extra includes
+#	$4: blob size
+# Note: _binary_$1_size is an "absolute" symbol. Because of relocation in
+# virtual memory systems (i.e. native) this will not work in such systems.
+# instead, we will have to hardcode the blob size.
+define blob_header_template
```
Comment (kaspar030) [inline makefiles/blobs.inc.mk]: Templating in a Makefile?

Why not implement the whole thing in a bash script?

File: makefiles/blobs.inc.mk [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ifeq (native,$(BOARD))
+  LD_BLOBFLAGS += -m elf_i386
+endif
+
+BLOB_SECTION ?= .rodata,alloc,load,readonly,data,contents
+OBJCOPY_RENAMEFLAGS ?= --rename-section .data=$(BLOB_SECTION)
+
+.INTERMEDIATE: $(BLOBS:%=$(BLOB_PATH)/%.o.tmp)
+
+$(BLOBS:%=$(BLOB_PATH)/%.o.tmp): $(BLOB_PATH)/%.o.tmp: % | $(BLOB_PATH)/
```
Comment (kaspar030) [inline makefiles/blobs.inc.mk]: Why use ld at all? objcopy can also create .o from blobs with a similar naming scheme.

**Comment by jcarrano** [comment]
> No arrays in headers.

If the array was in a C file, then OK. But it is in a header, meaning that:

- User must be aware that he should not include that header twice
- If he wants to use the array from different units, he must create a C file, include the array and create another h file with extern declarations.

As for creating a C file with an array, I don't remember the exact details, but I didn't play so nice with our build system.

The array approach does not allow one to embed code. Things like softdevice, for example, still require two images to be flashed.

**Comment by kaspar030** [comment]
> If the array was in a C file, then OK. But it is in a header, meaning that:

why not ```mv foo.c foo.h```?

**Comment by miri64** [comment]
Is @jcarrano aware of #11870. I'm not really sure what to make of that PR.

**Comment by aabadie** [comment]
#11870 was merged. Is this one still relevant ? I'm for closing it. @kaspar030 @miri64 what do you think ?

**Pull request closed without merging**